### ISSUE TITLE
Rule semicolon is behaving inconsistently/unexpectedly

### ISSUE THREAD

**Issue by mnn** [palantir/tslint#2015]
### Bug Report

- __TSLint version__: `4.3.1`
- __TypeScript version__: `2.1.4`
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
// using "semicolon": [true, "always"]
// actually doesnt mean it requires semicolon in all
// common positions (e.g. every field - just some fields).

class A {
  a = 0; // <- not reported
}

class B {
  b = () => ({b: 0}); // <- not reported
}

class C {
  c = () => ({
    c: 0
  }); // <- reported, WTH? It is same as a case B...
}

class D {
  d() {
    return 0;
  } // <- not reported (which is fine IMO)
}
```

with `tslint.json` configuration:

```json
...
    "semicolon": [true, "always"],
...
```

#### Actual behavior

C case is reported (`Unnecessary semicolon`).

#### Expected behavior

C case should not be reported. Or both fields with arrow function should be reported (B and C), but it would be IMO inconsistent, since other fields (case A) are not being reported.

**Comment by adidahiya** [comment]
See relevant discussion here: https://github.com/palantir/tslint/issues/1949#issuecomment-270809398

My assessment and (accepted) proposal:

> Right now the "always" option mostly just enforces the existence of semicolons in certain positions and only disallows semicolons in one position. The semantics could use some cleaning up; perhaps the reporting of "unnecessary" semicolons belongs in another rule option?

**Comment by bumbleblym** [comment]
the typescript spec https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md requires that

- type alias declarations are terminated with a semicolon
- object type members are separated by semicolons or commas; the last semicolon/comma is optional.

the compiler behaviour is considered(?) to be buggy: https://github.com/Microsoft/TypeScript/issues/10730

the semicolon rule applies to both cases, while the trailing-comma rule affects the second case: https://github.com/palantir/tslint/issues/1810

my proposal is to have semicolons/trailing-comma apply to statements and not type declarations, and add the following rule:

"object-type-member-separator": [true, "semicolon"]
"object-type-member-separator": [true, "comma"]
"object-type-member-separator": [true, "semicolon", no-final-separator"]

which would result in a breaking change.

**Comment by foxel** [comment]
What's the resolution here?

**Comment by ajafff** [comment]
#3294 added a `strict-bound-class-methods` option that was released with tslint 5.8.0
